Add an OOME check in ClonedArguments::createEmpty.

MenloDorian opened the pull request:
#### 102aa1b4c5015681b4339643f483a047cd7453c3
<pre>
Add an OOME check in ClonedArguments::createEmpty.
<a href="https://bugs.webkit.org/show_bug.cgi?id=274451">https://bugs.webkit.org/show_bug.cgi?id=274451</a>
<a href="https://rdar.apple.com/121951051">rdar://121951051</a>

Reviewed by Yusuke Suzuki.

Also added a RELEASE_ASSERT in operationCreateClonedArgumentsDuringExit() to make it explicit
that we&apos;re not allowed to have OOMEs (from ClonedArguments::createEmpty() or othewise) during
an OSR exit.  Encountering an OOME then is considered a non-recoverable event.

* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/ftl/FTLOperations.cpp:
(JSC::FTL::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/runtime/ClonedArguments.cpp:
(JSC::ClonedArguments::createEmpty):
(JSC::ClonedArguments::createWithInlineFrame):
(JSC::ClonedArguments::createByCopyingFrom):
* Source/JavaScriptCore/runtime/ClonedArguments.h:
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::createArrayStorageButterflyImpl):
(JSC::JSObject::createArrayStorageButterfly):
(JSC::JSObject::tryCreateArrayStorageButterfly):
* Source/JavaScriptCore/runtime/JSObject.h:

Canonical link: <a href="https://commits.webkit.org/279057@main">https://commits.webkit.org/279057@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b392e8b87a7506b7ee11a61961f622a05288e0ef

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52372 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31704 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4793 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55646 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3095 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38151 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2794 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/42590 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1975 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54468 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29338 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45202 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23671 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26558 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2477 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1254 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/45721 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48452 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2627 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57242 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/51880 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27498 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2656 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49977 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28731 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45322 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49226 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29643 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/64188 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7675 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28476 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12157 "Found 8 new JSC stress test failures: stress/map-iteration.js.dfg-eager, stress/proxy-set-prototype-of.js.mini-mode, stress/shadow-realm-remote-function-copy-length-and-name.js.dfg-eager, wasm.yaml/wasm/function-tests/brTableAsIf.js.wasm-eager, wasm.yaml/wasm/function-tests/f32-const.js.wasm-eager, wasm.yaml/wasm/function-tests/i32-load8-s.js.wasm-eager, wasm.yaml/wasm/function-tests/shl.js.wasm-eager, wasm.yaml/wasm/stress/mutable-globals.js.wasm-eager (failure)") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/ClonedArguments.cpp:
```diff
@@ -49,7 +49,11 @@ ClonedArguments* ClonedArguments::createEmpty(VM& vm, Structure* structure, JSFu
-            butterfly = createArrayStorageButterfly(vm, nullptr, structure, length, vectorLength);
+            butterfly = tryCreateArrayStorageButterfly(vm, nullptr, structure, length, vectorLength);
+            if (UNLIKELY(!butterfly)) {
+                throwOutOfMemoryError(structure->globalObject(), scope);
```
What happen when it is throwing an error from `operationCreateClonedArgumentsDuringExit` in OSR exit?

MenloDorian replied:
Hmmm, that is an interesting question.  However, it is a pre-existing condition as `ClonedArguments::createEmpty` was already able to throw an OOME before this.  See the else case.

I think an OOME during OSR exit is catastrophic, and we currently already crash when we try to put arguments into the ClonedArguments.  I'll add a RELEASE_ASSERT in `operationCreateClonedArgumentsDuringExit` to make it explicit that `ClonedArguments::createEmpty` must not return nullptr here.

Constellation commented on Source/JavaScriptCore/runtime/ClonedArguments.cpp:
```diff
@@ -49,7 +49,11 @@ ClonedArguments* ClonedArguments::createEmpty(VM& vm, Structure* structure, JSFu
 
     if (UNLIKELY(structure->mayInterceptIndexedAccesses() || structure->storedPrototypeObject()->needsSlowPutIndexing())) {
         if (!butterfly) {
-            butterfly = createArrayStorageButterfly(vm, nullptr, structure, length, vectorLength);
```
For `ClonedArguments::createEmpty` signature, let's take `JSGlobalObject*` as a first parameter. In general, we take that when the function throws an error.

MenloDorian replied:
Fixed.

Constellation reviewed (approved):
r=me with nit.

Commit:
Add an OOME check in ClonedArguments::createEmpty.
https://bugs.webkit.org/show_bug.cgi?id=274451
rdar://121951051

Reviewed by Yusuke Suzuki.

Also added a RELEASE_ASSERT in operationCreateClonedArgumentsDuringExit() to make it explicit
that we're not allowed to have OOMEs (from ClonedArguments::createEmpty() or othewise) during
an OSR exit.  Encountering an OOME then is considered a non-recoverable event.

* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/ftl/FTLOperations.cpp:
(JSC::FTL::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/runtime/ClonedArguments.cpp:
(JSC::ClonedArguments::createEmpty):
(JSC::ClonedArguments::createWithInlineFrame):
(JSC::ClonedArguments::createByCopyingFrom):
* Source/JavaScriptCore/runtime/ClonedArguments.h:
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::createArrayStorageButterflyImpl):
(JSC::JSObject::createArrayStorageButterfly):
(JSC::JSObject::tryCreateArrayStorageButterfly):
* Source/JavaScriptCore/runtime/JSObject.h:

Canonical link: https://commits.webkit.org/279057@main
